### ISSUE TITLE
Build error: use of undeclared identifier 'AUDIT_FILTER_EXCLUDE'

### ISSUE THREAD

**Issue by AE1020** [osquery/osquery#6551]
# Bug report

While in the step building libaudit there is a circular definition. It is caused because osquery libaudit submodule [is snapshotted at `20204cc`](https://github.com/osquery/osquery/tree/master/libraries/cmake/source/libaudit), which appears out of date. [In that version's libaudit.h](https://github.com/linux-audit/audit-userspace/blob/20204ccf43e856818d7ef695242192b3f5963aae/lib/libaudit.h#L263), it says:

    #define AUDIT_FILTER_EXCLUDE    AUDIT_FILTER_TYPE

But the definition of `AUDIT_FILTER_TYPE` on Kubuntu 20 in `<linux/audit.h>` is:

    #define AUDIT_FILTER_TYPE   AUDIT_FILTER_EXCLUDE /* obsolete misleading naming */
 
Clearly a problem. If the libaudit were brought up to [the version at time of writing](https://github.com/linux-audit/audit-userspace/blob/26922fd9b236754ebb5e993a4d5d37708545f916/lib/libaudit.h#L320), it would say:

    #ifndef AUDIT_FILTER_EXCLUDE
    #define AUDIT_FILTER_EXCLUDE    AUDIT_FILTER_TYPE

They seemed to [do this on purpose](https://www.redhat.com/archives/linux-audit/2018-June/msg00011.html).

### What operating system and version are you using?

    $ cat /etc/os-release
    NAME="Ubuntu"
    VERSION="20.04 LTS (Focal Fossa)"
    ID=ubuntu
    ID_LIKE=debian
    PRETTY_NAME="Ubuntu 20.04 LTS"
    VERSION_ID="20.04"
    VERSION_CODENAME=focal

    $ uname -r
    5.4.0-40-generic

I now notice [the build instructions](https://osquery.readthedocs.io/en/latest/development/building/) mention **Ubuntu 18.04/18.10** and not 20. But I was using Kubuntu 18.10 at first when trying to build osquery...but the CMake was too old, and 18.10 [just hit end of life](https://endoflife.software/operating-systems/linux/ubuntu#:~:text=Current%20releases%20%20%20%20Release%20%20,%20April%2C%202024%20%204%20more%20rows%20) (July 2019). So I upgraded.

### What version of osquery are you using?

Building current `git clone`.

### What steps did you take to reproduce the issue?

Out of curiosity I tried just commenting out the `#define AUDIT_FILTER_EXCLUDE` in libaudit.h, instead of updating the submodule. It got the compile to keep going, and maybe well enough to work.

### What did you expect to see?

Successful build. <sub>That is actually a lie. I never expect building a project to work the first time. :building_construction:</sub>

### What did you see instead?

Build failing.

**Comment by Smjert** [comment]
@AE1020 Thanks for the report!
This is something we are aware of unfortunately; we need to test if updating `libaudit` keeps the compatibility with the older platforms we support first.

A note about CMake being too old: in the instructions we purposedly make the user download and extract CMake from the official site so that the correct/updated version it's used.
The version they compile can run on a big range of distro versions, and even if one needs to compile it from scratch, it's very easy to do so, either with the already available CMake on the system (if recent enough, normally > 3.x) or with their bootstrap script.

**Comment by AE1020** [comment]
> A note about CMake being too old: in the instructions we purposedly make the user download and extract CMake from the official site so that the correct/updated version it's used.

Yes I could have. But I noticed a lot of other errors tied into the end-of-life in the apt install for the VM I pulled up. So I decided it would be a good time to upgrade, and I'd get the new CMake while I was at it. Famous last words.

**Comment by directionless** [comment]
Thanks for the detailed analysis of the circular macro definition between the bundled libaudit and kernel headers — that was a very clear writeup.

Since this was filed, osquery's third-party library dependencies have been updated, and the libaudit submodule has been bumped past the commit that contained this issue. The upstream fix (wrapping the define in `#ifndef AUDIT_FILTER_EXCLUDE`) is included in newer versions.

Closing this as resolved by subsequent dependency updates. If you're still hitting build issues on a current checkout, please open a new issue with the updated details.
